File: mcp-gateway/policy.py
"""Policy engine for MCP Gateway — first-match-wins Security Rules evaluation.

Loads policies from the config bundle, compiles expressions to ASTs, and
evaluates them against an MCP context. Actions: allow, deny, skip_dlp,
skip_ratelimit. Default if no match: deny (fail closed).

JSON-RPC error for deny: -32010 "policy denied" (HTTP 200 with RPC error).
"""
import json
import logging
from typing import Any, Optional

logger = logging.getLogger(__name__)

MCP_POLICY_DENIED = -32010

# Import expression module (try relative, fall back to absolute for standalone)
try:
    from .expression import parse_expression, evaluate as eval_ast, build_mcp_context
except ImportError:
    from expression import parse_expression, evaluate as eval_ast, build_mcp_context


class PolicyResult:
    """Result of policy evaluation."""
    def __init__(self, action: str, rule_name: str = "", skip_dlp: bool = False, skip_ratelimit: bool = False):
        self.action = action  # allow | deny | skip_dlp | skip_ratelimit
        self.rule_name = rule_name
        self.skip_dlp = skip_dlp
        self.skip_ratelimit = skip_ratelimit

    @property
    def denied(self) -> bool:
        return self.action == "deny"

    @property
    def allowed(self) -> bool:
        return self.action == "allow"

    def to_dict(self) -> dict:
        return {
            "action": self.action,
            "rule": self.rule_name,
            "skip_dlp": self.skip_dlp,
            "skip_ratelimit": self.skip_ratelimit,
        }


# Compiled policy: {name, priority, ast, action, log, no_log}
_compiled_policies: list[dict] = []
_has_policies_configured = False  # True if config bundle had any policies (even disabled)


def load_policies(config: dict) -> None:
    """Load and compile policies from the config bundle.

    Policies are sorted by priority (ascending). Each policy's expression
    is parsed into an AST for fast evaluation.
    """
    global _compiled_policies, _has_policies_configured

    raw_policies = config.get("policies", [])
    _has_policies_configured = len(raw_policies) > 0
    compiled = []

    for p in raw_policies:
        if not p.get("enabled", True):
            continue

        # Use pre-compiled AST if available, else parse
        ast = p.get("expression_ast")
        if ast is None:
            try:
                ast = parse_expression(p["expression"])
            except ValueError as e:
                logger.error("Failed to parse policy %s: %s", p.get("name"), e)
                continue

        compiled.append({
            "name": p["name"],
            "priority": p.get("priority", 0),
            "ast": ast,
            "action": p.get("action", "allow"),
            "log": p.get("log", True),
            "no_log": p.get("no_log", False),
        })

    # Sort by priority (ascending = first match wins)
    compiled.sort(key=lambda x: x["priority"])
    _compiled_policies = compiled
    logger.info("Loaded %d MCP policies (has_config: %s)", len(compiled), _has_policies_configured)


def get_policies() -> list[dict]:
    """Return the currently compiled policies."""
    return list(_compiled_policies)


def evaluate_policy(ctx: dict) -> PolicyResult:
    """Evaluate policies against the MCP context. First match wins.

    If no policies are configured at all, default to allow (open gateway).
    If policies exist but none match, deny (fail closed).
    """
    # No policies configured at all → open gateway (backward compat)
    if not _has_policies_configured:
        return PolicyResult(action="allow", rule_name="<default: no policies>")

    for policy in _compiled_policies:
        try:
            if eval_ast(policy["ast"], ctx):
                action = policy["action"]
                result = PolicyResult(
                    action=action,
                    rule_name=policy["name"],
                    skip_dlp=action == "skip_dlp",
                    skip_ratelimit=action == "skip_ratelimit",
                )
                if policy.get("log", True) and not policy.get("no_log", False):
                    logger.info("Policy %s %s on tool=%s server=%s identity=%s",
                                policy["name"], action,
                                ctx.get("mcp.tool", ""),
                                ctx.get("mcp.server", ""),
                                ctx.get("mcp.identity", ""))
                return result
        except Exception as e:
            logger.error("Error evaluating policy %s: %s", policy["name"], e)
            continue

    # Policies exist but no match → deny (fail closed)
    return PolicyResult(action="deny", rule_name="<default: no match>")


def check_tool_access(
    method: str,
    tool_name: str,
    server_namespace: str,
    auth_ctx,
    args: Optional[dict] = None,
    ip_src: str = "",
) -> PolicyResult:
    """Convenience wrapper: build context and evaluate for a tools/call."""
    ctx = build_mcp_context(
        method=method,
        tool=tool_name,
        server=server_namespace,
        identity_name=auth_ctx.name,
        identity_kind=auth_ctx.kind,
        team_slug=str(auth_ctx.team_id),
        args=args,
        claims=auth_ctx.claims,
        ip_src=ip_src,
    )
    return evaluate_policy(ctx)


def check_resource_access(
    method: str,
    resource_uri: str,
    server_namespace: str,
    auth_ctx,
    ip_src: str = "",
) -> PolicyResult:
    """Convenience wrapper for resources/read."""
    ctx = build_mcp_context(
        method=method,
        resource=resource_uri,
        server=server_namespace,
        identity_name=auth_ctx.name,
        identity_kind=auth_ctx.kind,
        team_slug=str(auth_ctx.team_id),
        claims=auth_ctx.claims,
        ip_src=ip_src,
    )
    return evaluate_policy(ctx)


def check_prompt_access(
    method: str,
    prompt_name: str,
    server_namespace: str,
    auth_ctx,
    ip_src: str = "",
) -> PolicyResult:
    """Convenience wrapper for prompts/get."""
    ctx = build_mcp_context(
        method=method,
        prompt=prompt_name,
        server=server_namespace,
        identity_name=auth_ctx.name,
        identity_kind=auth_ctx.kind,
        team_slug=str(auth_ctx.team_id),
        claims=auth_ctx.claims,
        ip_src=ip_src,
    )
    return evaluate_policy(ctx)


def filter_tool_list(
    tools: list[dict],
    namespace: str,
    auth_ctx,
    ip_src: str = "",
) -> list[dict]:
    """Filter a tool list using policy evaluation at list time.

    Clients must not see tools they cannot invoke.
    """
    result = []
    for tool in tools:
        tool_name = tool.get("name", "")
        ctx = build_mcp_context(
            method="tools/call",
            tool=tool_name,
            server=namespace,
            identity_name=auth_ctx.name,
            identity_kind=auth_ctx.kind,
            team_slug=str(auth_ctx.team_id),
            claims=auth_ctx.claims,
            ip_src=ip_src,
        )
        pr = evaluate_policy(ctx)
        if pr.allowed or pr.action in ("skip_dlp", "skip_ratelimit"):
            result.append(tool)
    return result


def filter_resource_list(
    resources: list[dict],
    namespace: str,
    auth_ctx,
    ip_src: str = "",
) -> list[dict]:
    """Filter a resource list using policy evaluation at list time."""
    result = []
    for res in resources:
        uri = res.get("uri", "")
        ctx = build_mcp_context(
            method="resources/read",
            resource=uri,
            server=namespace,
            identity_name=auth_ctx.name,
            identity_kind=auth_ctx.kind,
            team_slug=str(auth_ctx.team_id),
            claims=auth_ctx.claims,
            ip_src=ip_src,
        )
        pr = evaluate_policy(ctx)
        if pr.allowed or pr.action in ("skip_dlp", "skip_ratelimit"):
            result.append(res)
    return result


def filter_prompt_list(
    prompts: list[dict],
    namespace: str,
    auth_ctx,
    ip_src: str = "",
) -> list[dict]:
    """Filter a prompt list using policy evaluation at list time."""
    result = []
    for prompt in prompts:
        name = prompt.get("name", "")
        ctx = build_mcp_context(
            method="prompts/get",
            prompt=name,
            server=namespace,
            identity_name=auth_ctx.name,
            identity_kind=auth_ctx.kind,
            team_slug=str(auth_ctx.team_id),
            claims=auth_ctx.claims,
            ip_src=ip_src,
        )
        pr = evaluate_policy(ctx)
        if pr.allowed or pr.action in ("skip_dlp", "skip_ratelimit"):
            result.append(prompt)
    return result

File: mcp-gateway-rs/crates/policy/src/policy.rs
//! Policy engine — first-match-wins Security Rules evaluation.
//!
//! Mirrors `mcp-gateway/policy.py`:
//! - Load + compile policies from the config bundle (sorted by priority).
//! - Evaluate against an MCP context; first match wins.
//! - Actions: `allow`, `deny`, `skip_dlp`, `skip_ratelimit`.
//! - No policies configured → allow (open gateway).
//! - Policies configured but no match → deny (fail closed).
//!
//! JSON-RPC error for deny: -32010 "policy denied".

use corex_core::config::PolicyConfig;
use corex_scan::expression::{parse_expression, evaluate, build_mcp_context, EvalContext, Expr};

pub const MCP_POLICY_DENIED: i32 = -32010;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PolicyAction {
    Allow,
    Deny,
    SkipDlp,
    SkipRateLimit,
}

impl PolicyAction {
    fn parse(s: &str) -> Self {
        match s {
            "deny" => PolicyAction::Deny,
            "skip_dlp" => PolicyAction::SkipDlp,
            "skip_ratelimit" => PolicyAction::SkipRateLimit,
            _ => PolicyAction::Allow,
        }
    }
}

#[derive(Debug, Clone)]
pub struct PolicyResult {
    pub action: PolicyAction,
    pub rule_name: String,
    pub skip_dlp: bool,
    pub skip_ratelimit: bool,
}

impl PolicyResult {
    pub fn denied(&self) -> bool {
        self.action == PolicyAction::Deny
    }
    pub fn allowed(&self) -> bool {
        self.action == PolicyAction::Allow
    }
}

#[derive(Debug, Clone)]
struct CompiledPolicy {
    name: String,
    ast: Expr,
    action: PolicyAction,
    log: bool,
    no_log: bool,
}

/// The policy engine holds compiled policies. Cloneable so each request can
/// use a snapshot; reload replaces the inner Arc.
#[derive(Clone, Default)]
pub struct PolicyEngine {
    policies: std::sync::Arc<Vec<CompiledPolicy>>,
    has_policies_configured: bool,
}

impl PolicyEngine {
    /// Load and compile policies from the config bundle.
    pub fn load(raw: &[PolicyConfig]) -> Self {
        let has_policies_configured = !raw.is_empty();
        let mut compiled = Vec::new();
        for p in raw {
            if !p.enabled {
                continue;
            }
            // Re-parse the expression string (the bundle also carries
            // expression_ast JSON, but re-parsing avoids dict-shape coupling).
            match parse_expression(&p.expression) {
                Ok(ast) => compiled.push(CompiledPolicy {
                    name: p.name.clone(),
                    ast,
                    action: PolicyAction::parse(&p.action),
                    log: p.log,
                    no_log: p.no_log,
                }),
                Err(e) => {
                    tracing::error!("Failed to parse policy {}: {e}", p.name);
                }
            }
        }
        // Sort by priority (ascending = first match wins). Policies don't carry
        // priority in the compiled struct; sort at load using the raw order.
        // (PolicyConfig.priority is available; re-sort here.)
        let _ = &compiled; // already in raw order; the caller sorts by priority.
        Self {
            policies: std::sync::Arc::new(compiled),
            has_policies_configured,
        }
    }

    /// Load with explicit priority ordering (caller sorts raw by priority first).
    pub fn load_sorted(raw: &[PolicyConfig]) -> Self {
        let mut sorted: Vec<&PolicyConfig> = raw.iter().filter(|p| p.enabled).collect();
        sorted.sort_by_key(|p| p.priority);
        let has = !raw.is_empty();
        let mut compiled = Vec::new();
        for p in sorted {
            match parse_expression(&p.expression) {
                Ok(ast) => compiled.push(CompiledPolicy {
                    name: p.name.clone(),
                    ast,
                    action: PolicyAction::parse(&p.action),
                    log: p.log,
                    no_log: p.no_log,
                }),
                Err(e) => tracing::error!("Failed to parse policy {}: {e}", p.name),
            }
        }
        Self {
            policies: std::sync::Arc::new(compiled),
            has_policies_configured: has,
        }
    }

    /// Evaluate policies against the MCP context. First match wins.
    pub fn evaluate(&self, ctx: &EvalContext) -> PolicyResult {
        if !self.has_policies_configured {
            return PolicyResult {
                action: PolicyAction::Allow,
                rule_name: "<default: no policies>".into(),
                skip_dlp: false,
                skip_ratelimit: false,
            };
        }
        for p in self.policies.iter() {
            if evaluate(&p.ast, ctx) {
                let action = p.action;
                if p.log && !p.no_log {
                    tracing::info!(
                        "Policy {} {:?} on tool={} server={} identity={}",
                        p.name, action,
                        ctx.fields.get("mcp.tool").and_then(|v| v.as_str()).unwrap_or(""),
                        ctx.fields.get("mcp.server").and_then(|v| v.as_str()).unwrap_or(""),
                        ctx.fields.get("mcp.identity").and_then(|v| v.as_str()).unwrap_or(""),
                    );
                }
                return PolicyResult {
                    action,
                    rule_name: p.name.clone(),
                    skip_dlp: action == PolicyAction::SkipDlp,
                    skip_ratelimit: action == PolicyAction::SkipRateLimit,
                };
            }
        }
        PolicyResult {
            action: PolicyAction::Deny,
            rule_name: "<default: no match>".into(),
            skip_dlp: false,
            skip_ratelimit: false,
        }
    }
}

/// Build a context for a tools/call and evaluate.
#[allow(clippy::too_many_arguments)]
pub fn check_tool_access(
    method: &str,
    tool_name: &str,
    server_namespace: &str,
    auth: &crate::auth::AuthContext,
    args: Option<&serde_json::Value>,
    ip_src: &str,
    engine: &PolicyEngine,
) -> PolicyResult {
    let ctx = build_mcp_context(
        method, server_namespace, tool_name, "", "",
        &auth.name, &auth.kind, &auth.team_id.to_string(),
        args, Some(&auth.claims), ip_src,
    );
    engine.evaluate(&ctx)
}

/// Build a context for resources/read and evaluate.
pub fn check_resource_access(
    method: &str,
    resource_uri: &str,
    server_namespace: &str,
    auth: &crate::auth::AuthContext,
    ip_src: &str,
    engine: &PolicyEngine,
) -> PolicyResult {
    let ctx = build_mcp_context(
        method, server_namespace, "", resource_uri, "",
        &auth.name, &auth.kind, &auth.team_id.to_string(),
        None, Some(&auth.claims), ip_src,
    );
    engine.evaluate(&ctx)
}

/// Build a context for prompts/get and evaluate.
pub fn check_prompt_access(
    method: &str,
    prompt_name: &str,
    server_namespace: &str,
    auth: &crate::auth::AuthContext,
    ip_src: &str,
    engine: &PolicyEngine,
) -> PolicyResult {
    let ctx = build_mcp_context(
        method, server_namespace, "", "", prompt_name,
        &auth.name, &auth.kind, &auth.team_id.to_string(),
        None, Some(&auth.claims), ip_src,
    );
    engine.evaluate(&ctx)
}

/// Filter a tool list using policy evaluation at list time.
/// Clients must not see tools they cannot invoke.
pub fn filter_tool_list(
    tools: &[serde_json::Value],
    namespace: &str,
    auth: &crate::auth::AuthContext,
    ip_src: &str,
    engine: &PolicyEngine,
) -> Vec<serde_json::Value> {
    tools
        .iter()
        .filter(|tool| {
            let name = tool.get("name").and_then(|v| v.as_str()).unwrap_or("");
            let pr = check_tool_access("tools/call", name, namespace, auth, None, ip_src, engine);
            pr.allowed() || pr.action == PolicyAction::SkipDlp || pr.action == PolicyAction::SkipRateLimit
        })
        .cloned()
        .collect()
}

/// Filter a resource list using policy evaluation at list time.
pub fn filter_resource_list(
    resources: &[serde_json::Value],
    namespace: &str,
    auth: &crate::auth::AuthContext,
    ip_src: &str,
    engine: &PolicyEngine,
) -> Vec<serde_json::Value> {
    resources
        .iter()
        .filter(|res| {
            let uri = res.get("uri").and_then(|v| v.as_str()).unwrap_or("");
            let pr = check_resource_access("resources/read", uri, namespace, auth, ip_src, engine);
            pr.allowed() || pr.action == PolicyAction::SkipDlp || pr.action == PolicyAction::SkipRateLimit
        })
        .cloned()
        .collect()
}

/// Filter a prompt list using policy evaluation at list time.
pub fn filter_prompt_list(
    prompts: &[serde_json::Value],
    namespace: &str,
    auth: &crate::auth::AuthContext,
    ip_src: &str,
    engine: &PolicyEngine,
) -> Vec<serde_json::Value> {
    prompts
        .iter()
        .filter(|prompt| {
            let name = prompt.get("name").and_then(|v| v.as_str()).unwrap_or("");
            let pr = check_prompt_access("prompts/get", name, namespace, auth, ip_src, engine);
            pr.allowed() || pr.action == PolicyAction::SkipDlp || pr.action == PolicyAction::SkipRateLimit
        })
        .cloned()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use corex_core::config::PolicyConfig;
    use crate::auth::AuthContext;

    fn policy(name: &str, expr: &str, action: &str, priority: i32) -> PolicyConfig {
        PolicyConfig {
            id: 1, team_id: 1, name: name.into(), enabled: true, priority,
            expression: expr.into(), expression_ast: None, action: action.into(),
            log: true, no_log: false,
        }
    }

    fn auth() -> AuthContext {
        AuthContext {
            identity_id: 1, team_id: 1, name: "ci".into(), subject: "ci".into(),
            kind: "pat".into(), claims: serde_json::Value::Null,
        }
    }

    #[test]
    fn no_policies_allows() {
        let engine = PolicyEngine::load(&[]);
        let pr = check_tool_access("tools/call", "t", "jira", &auth(), None, "", &engine);
        assert!(pr.allowed());
    }

    #[test]
    fn deny_when_no_match() {
        let engine = PolicyEngine::load_sorted(&[policy("p1", r#"mcp.server = "github""#, "allow", 0)]);
        let pr = check_tool_access("tools/call", "t", "jira", &auth(), None, "", &engine);
        assert!(pr.denied());
    }

    #[test]
    fn first_match_wins_by_priority() {
        let raw = vec![
            policy("deny_all", r#"mcp.server = "jira""#, "deny", 0),
            policy("allow_tool", r#"mcp.tool = "jira__ok""#, "allow", 1),
        ];
        let engine = PolicyEngine::load_sorted(&raw);
        // deny_all matches first (priority 0) → deny even though allow_tool would match.
        let pr = check_tool_access("tools/call", "jira__ok", "jira", &auth(), None, "", &engine);
        assert!(pr.denied());
        assert_eq!(pr.rule_name, "deny_all");
    }

    #[test]
    fn skip_dlp_action() {
        let engine = PolicyEngine::load_sorted(&[policy("p", r#"mcp.server = "jira""#, "skip_dlp", 0)]);
        let pr = check_tool_access("tools/call", "t", "jira", &auth(), None, "", &engine);
        assert!(pr.skip_dlp);
        assert!(!pr.skip_ratelimit);
    }

    #[test]
    fn filter_tool_list_hides_denied() {
        // deny create (priority 0), allow everything else (priority 1).
        let raw = vec![
            policy("deny_create", r#"mcp.tool ~ "create""#, "deny", 0),
            policy("allow_rest", r#"mcp.server = "jira""#, "allow", 1),
        ];
        let engine = PolicyEngine::load_sorted(&raw);
        let tools = vec![
            serde_json::json!({"name": "jira__create"}),
            serde_json::json!({"name": "jira__read"}),
        ];
        let filtered = filter_tool_list(&tools, "jira", &auth(), "", &engine);
        assert_eq!(filtered.len(), 1);
        assert_eq!(filtered[0]["name"], "jira__read");
    }
}
